packages/appsync-modelgen-plugin/src/utils/process-has-many.ts

```diff
@@ -134,3 +135,82 @@ export function getConnectedFieldsForHasMany(
       }
     });
 }
+
+/**
+ * Helper to add a key directive to a given model.
+ */
+ function addKeyToModel(model: CodeGenModel, name: string, fields: string[]): void {
+  model.directives.push({
+    name: 'key',
+    arguments: {
+      name,
+      fields,
+    },
+  });
+}
+
+/**
+ * Helper to simplify retrieving either CPK-enabled or standard connection fields.
+ * Returns a list of fields of at least length 1.
+ */
+function getConnectionAssociatedFields(hasManyConnection: CodeGenFieldConnectionHasMany): CodeGenField[] {
+  const associatedFields = hasManyConnection.associatedWithFields && hasManyConnection.associatedWithFields.length > 0
+  ? hasManyConnection.associatedWithFields
+  : [hasManyConnection.associatedWith]
+  if (associatedFields.length === 0) {
+    throw new Error('Expected at least one associated field for the hasMany relationship.');
+  }
+  return associatedFields;
+}
+
+/**
+ * A corresponding belongsTo means that the corresponding model has a field with type that is the name of this model, and with a @belongsTo
+ * directive attached to it.
+ */
+function doesHasManyConnectionHaveCorrespondingBelongsTo(model: CodeGenModel, hasManyConnection: CodeGenFieldConnectionHasMany): boolean {
+  const fieldReferencingParent = hasManyConnection.connectedModel.fields.find(f => f.type === model.name);
+  if (!fieldReferencingParent) {
+    return false;
+  }
+  return fieldReferencingParent.directives.some(d => d.name === TransformerV2DirectiveName.BELONGS_TO);
+}
+
+/**
+ * Couple of helper types to make the filtering logic below more concise.
+ */
+type FieldFilter = (f: CodeGenField) => boolean;
+type DirectiveFiter = (d: CodeGenDirective) => boolean;
+
+/**
+ * This is a bit backwards, but we need to check if this connection directive specifies an index.
+ */
+function doesHasManySpecifyIndexName(model: CodeGenModel, hasManyConnection: CodeGenFieldConnectionHasMany): boolean {
+  const directiveIsHasMany: DirectiveFiter = (d) => d.name === TransformerV2DirectiveName.HAS_MANY;
+  const directiveHasIndexName: DirectiveFiter = (d) => d.arguments.indexName;
+  const fieldHasHasManyDirectiveWithIndex: FieldFilter = (f) => f.directives.some(d => directiveIsHasMany(d) && directiveHasIndexName(d));
+  const fieldMatchesConnectedModelName: FieldFilter = (f) => f.type === hasManyConnection.connectedModel.name;
+  return model.fields.some(f => fieldMatchesConnectedModelName(f) && fieldHasHasManyDirectiveWithIndex(f));
```

Comment:
This will probably return false positive if there are multiple `hasMany` with the same child type. Still I am doubt about its accuracy of this kind of schema. 
```GraqhQL
type Post @model {
  comments1: [Comment] @hasMany
  comments2: [Comment] @hasMany(indexName: "byComment2")
}
type Comment @model {
  content: String
}
```

Reply:
One solution is that instead of finding fields we can directly use the field from the loop as mentioned in the base visitor

Reply:
Makes sense, yeah leveraging the field from the visitor should mitigate the need for this matcher.